src/proto/h2/mod.rs

```diff
@@ -243,32 +242,201 @@ impl<B: Buf> SendStreamExt for SendStream<SendBuf<B>> {
+        let cnt = std::cmp::min(self.buf.len(), read_buf.remaining());
+        read_buf.put_slice(&self.buf[..cnt]);
+        self.buf.advance(cnt);
+        let _ = self.recv_stream.flow_control().release_capacity(cnt);
+        Poll::Ready(Ok(()))
+    }
+}
+
+impl<B> AsyncWrite for H2Upgraded<B>
+where
+    B: Buf,
+{
+    fn poll_write(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &[u8],
+    ) -> Poll<Result<usize, io::Error>> {
+        if let Poll::Ready(reset) = self.send_stream.poll_reset(cx) {
+            return Poll::Ready(Err(h2_to_io_error(match reset {
+                Ok(reason) => reason.into(),
+                Err(e) => e,
+            })));
+        }
+        if buf.is_empty() {
+            return Poll::Ready(Ok(0));
+        }
+        self.send_stream.reserve_capacity(buf.len());
+        Poll::Ready(match ready!(self.send_stream.poll_capacity(cx)) {
+            None => Ok(0),
+            Some(Ok(cnt)) => self.send_stream.write(&buf[..cnt], false).map(|()| cnt),
+            Some(Err(e)) => Err(h2_to_io_error(e)),
+        })
+    }
+
+    fn poll_flush(self: Pin<&mut Self>, _cx: &mut Context<'_>) -> Poll<Result<(), io::Error>> {
+        Poll::Ready(Ok(()))
+    }
+
+    fn poll_shutdown(
+        mut self: Pin<&mut Self>,
+        _cx: &mut Context<'_>,
+    ) -> Poll<Result<(), io::Error>> {
+        Poll::Ready(self.send_stream.write(&[], true))
+    }
+}
+
+fn h2_to_io_error(e: h2::Error) -> io::Error {
+    if e.is_io() {
+        e.into_io().unwrap()
+    } else {
+        io::Error::new(io::ErrorKind::Other, e)
+    }
+}
+
+struct UpgradedSendStream<B>(SendStream<SendBuf<Neutered<B>>>);
+
+impl<B> UpgradedSendStream<B>
+where
+    B: Buf,
+{
+    unsafe fn new(inner: SendStream<SendBuf<B>>) -> Self {
+        assert_eq!(mem::size_of::<B>(), mem::size_of::<Neutered<B>>());
+        Self(mem::transmute(inner))
+    }
+
+    fn reserve_capacity(&mut self, cnt: usize) {
+        unsafe { self.as_inner_unchecked().reserve_capacity(cnt) }
+    }
+
+    fn poll_capacity(&mut self, cx: &mut Context<'_>) -> Poll<Option<Result<usize, h2::Error>>> {
+        unsafe { self.as_inner_unchecked().poll_capacity(cx) }
+    }
+
+    fn poll_reset(&mut self, cx: &mut Context<'_>) -> Poll<Result<h2::Reason, h2::Error>> {
+        unsafe { self.as_inner_unchecked().poll_reset(cx) }
+    }
+
+    fn write(&mut self, buf: &[u8], end_of_stream: bool) -> Result<(), io::Error> {
+        let send_buf = SendBuf::Cursor(Cursor::new(buf.into()));
+        unsafe {
+            self.as_inner_unchecked()
+                .send_data(send_buf, end_of_stream)
+                .map_err(h2_to_io_error)
+        }
+    }
+
+    unsafe fn as_inner_unchecked(&mut self) -> &mut SendStream<SendBuf<B>> {
+        &mut *(&mut self.0 as *mut _ as *mut _)
+    }
+}
+
+#[repr(transparent)]
+struct Neutered<B> {
+    _inner: B,
+    impossible: Impossible,
+}
+
+enum Impossible {}
+
+unsafe impl<B> Send for Neutered<B> {}
```

Comment:
This is my cursed idea to remove the `Send` bounds on `B::Data`. We know we are not ever actually going to send a `B::Data` at any point over that stream once it is upgraded, so AFAIK it doesn't matter whether it is `Send` or not.

So instead of introducing that `Send` bound, we transmute `SendStream<SendBuf<B>>` to `SendStream<SendBuf<Neutered<B>>>` where `Neutered<B>` forcefully implements `Send` because it cannot actually be fully initialised, as one of its fields is an empty type.

`Neutered<B>` though still needs to implement `Buf` because the definition of `SendStream<B>` itself requires `B: Buf` (that could probably be relaxed in h2, now that I think about it, I personally dislike unnecessary trait bounds on type definitions, they just make the code more verbose).

Note that we don't actually use the `SendStream<SendBuf<Neutered<B>>>`, instead we use `UpgradedSendStream::as_inner_unchecked` to retrieve the `SendStream<SendBuf<B>>` to minimize the amounts of monomorphisations for `SendStream<_>` methods.

Reply:
This is quite the monster. I do think it's clear what it does (maybe a comment on the `Neutered` type and/or `impossible` field would be good). It's clever, I think I like it!

My one question is if the transmute is safe. I know the empty enum has no size, but could optimizations in the compiler take advantage of it being impossible to create that type, and decide "it's impossible to create `SendBuf::Buf(Neutered)`, so we'll just chop that variant out completely"? I'd prefer if we find the answer, we could include a code comment pointing to that guarantee.

Reply:
> My one question is if the transmute is safe. I know the empty enum has no size, but could optimizations in the compiler take advantage of it being impossible to create that type, and decide "it's impossible to create SendBuf::Buf(Neutered), so we'll just chop that variant out completely"? I'd prefer if we find the answer, we could include a code comment pointing to that guarantee.

I'll add an assertion, but AFAIK rustc never kills variants that have *some* impossible fields but have others that are not zero-sized.

Reply:
Wait I already check this, that's what the assertion in `UpgradedSendStream::new` verifies.

Reply:
I've added `#[repr(usize)]` on `SendBuf<B>` to be extra sure rustc doesn't get fancy about the representation of the enum.